### PR TITLE
Implemented Item Loss Chance

### PR DESCRIPTION
**Pull request by Faun471**

### CODE REVIEW

File: src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java [hunk 1]
```diff
@@ -18,17 +23,65 @@ public void onPlayerMove(PlayerMoveEvent event) {
...
                     }
+
+                    ItemStack originalItem = item.clone();
+
+                    for (int i = 0; i < item.getAmount(); i++) {
+                        if (Math.random() * 100 > voidEnhancementData.itemLossChance) {
+                            continue;
+                        }
+
+                        lostItemsMap.put(originalItem, lostItemsMap.getOrDefault(originalItem, 0) + 1);
```
Comment (MertUnverdi) [inline src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java:59]: Use Map#compute instead

File: src/main/java/com/iridium/iridiumskyblock/database/LostItems.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+@DatabaseTable(tableName = "lost_items")
+public class LostItems extends DatabaseObject {
+    @DatabaseField(columnName = "id", generatedId = true, canBeNull = false, unique = true)
+    private int id;
+
+    @DatabaseField(columnName = "uuid", canBeNull = false)
+    private @NotNull UUID uuid;
+
+    @DatabaseField(columnName = "time")
+    private long time;
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/database/LostItems.java:29]: Sorry, I forgot to tell you...
https://github.com/Iridium-Development/IridiumTeams/blob/master/src/main/java/com/iridium/iridiumteams/database/types/LocalDateTimeType.java
We have automatic conversions of Datetime, so you can just do

```suggestion
    private LocalDateTime time;
```

File: src/main/java/com/iridium/iridiumskyblock/database/LostItems.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    @DatabaseField(columnName = "items", dataType = DataType.SERIALIZABLE)
+    private String[] items;
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/database/LostItems.java:32]: Have you tested this? didnt know you could have an array here?

File: src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java [hunk 4]
```diff
@@ -18,17 +23,65 @@ public void onPlayerMove(PlayerMoveEvent event) {
...
 
         user.getCurrentIsland().ifPresent(island -> {
-            if (event.getPlayer().getLocation().getY() < LocationUtils.getMinHeight(event.getPlayer().getWorld())) {
-                VoidEnhancementData voidEnhancementData = IridiumSkyblock.getInstance().getEnhancements().voidEnhancement.levels.get(IridiumSkyblock.getInstance().getTeamManager().getTeamEnhancement(island, "void").getLevel());
-                if (voidEnhancementData != null && voidEnhancementData.enabled) {
-                    if (IridiumSkyblock.getInstance().getTeamManager().teleport(event.getPlayer(), island.getHome(), island)) {
-                        event.getPlayer().sendMessage(StringUtils.color(IridiumSkyblock.getInstance().getMessages().voidTeleport
-                                .replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix)
-                        ));
+            if (!(event.getPlayer().getLocation().getY() < LocationUtils.getMinHeight(event.getPlayer().getWorld()))) {
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java:26]:
```suggestion
            if (event.getPlayer().getLocation().getY() >= LocationUtils.getMinHeight(event.getPlayer().getWorld())) {
```

File: src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java [hunk 5]
```diff
@@ -18,17 +23,65 @@ public void onPlayerMove(PlayerMoveEvent event) {
...
+
+            if (!IridiumSkyblock.getInstance().getTeamManager().teleport(event.getPlayer(), island.getHome(), island)) {
+                return;
+            }
+
+            event.getPlayer().sendMessage(StringUtils.color(IridiumSkyblock.getInstance().getMessages().voidTeleport
+                    .replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix)));
+
+            HashMap<ItemStack, Integer> lostItemsMap = new HashMap<>();
+            if (voidEnhancementData.itemLossChance > 0) {
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java:46]: We should use early return pattern
```suggestion
            if (voidEnhancementData.itemLossChance <= 0) return
```

File: src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java [hunk 6]
```diff
@@ -18,17 +23,65 @@ public void onPlayerMove(PlayerMoveEvent event) {
...
+
+                    for (int i = 0; i < item.getAmount(); i++) {
+                        if (Math.random() * 100 > voidEnhancementData.itemLossChance) {
+                            continue;
+                        }
+
+                        lostItemsMap.compute(originalItem, (key, value) -> value == null ? 1 : value + 1);
+                    }
+
+                    item.setAmount(item.getAmount() - lostItemsMap.getOrDefault(originalItem, 0));
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java:62]: Sorry, it took me a while to figure out what this was doing...

Would this make more sence?

```suggestion
                    ArrayList<ItemStack> lostItems = new ArrayList<>();
                    ItemStack originalItem = item.clone();
                    int lostAmount = 0;

                    for (int i = 0; i < item.getAmount(); i++) {
                        if (Math.random() * 100 <= voidEnhancementData.itemLossChance) {
                            lostAmount++;
                        }
                    }
                    if(lostAmount == 0) continue;
                    origionalItem.setAmount(lostAmount);
                    lostItemsMap.add(origionalItem);

                    int newAmount = originalItem.getAmount() - lostAmount;
                    item.setAmount(newAmount);
```

File: src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java [hunk 7]
```diff
@@ -18,17 +23,65 @@ public void onPlayerMove(PlayerMoveEvent event) {
...
                 }
+
+                if (lostItemsMap.isEmpty()) {
+                    return;
+                }
+
+                IridiumSkyblock.getInstance().getDatabaseManager().getLostItemsTableManager().addEntry(new LostItems(
+                        event.getPlayer().getUniqueId(), lostItemsMap.keySet().toArray(new ItemStack[0])));
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java:70]: Should store this as an array, also the amount of this Item should reflect the amount lost, not the origional amount

File: src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java [hunk 8]
```diff
@@ -18,17 +23,65 @@ public void onPlayerMove(PlayerMoveEvent event) {
...
+                IridiumSkyblock.getInstance().getDatabaseManager().getLostItemsTableManager().save();
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java:71]: TableManager#save does sql database operations syncronously, so for performance reasons we should do this async.

If reccomend moving this to the IridiumSkyblock#save function

File: src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java [hunk 9]
```diff
@@ -18,17 +23,65 @@ public void onPlayerMove(PlayerMoveEvent event) {
...
+                        .getMessages().voidLostItems
+                        .replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix)
+                        .replace("%items%", String.join(", ", Arrays
+                                .stream(lostItemsMap.keySet().toArray(new ItemStack[0]))
+                                .map(item -> IridiumSkyblock.getInstance().getMessages().itemsString
+                                        .replace("%amount%", lostItemsMap.get(item).toString())
+                                        .replace("%item_name%", item.getItemMeta().hasDisplayName() ? item.getItemMeta().getDisplayName() : "%type%")
+                                        .replace("%type%", item.getType().name().trim().replace("_", " ")))
+                                .toArray(String[]::new)))
+                ));
```
Comment (PeachesMLG) [inline src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java:83]:
```suggestion
                event.getPlayer().sendMessage(StringUtils.color(IridiumSkyblock.getInstance()
                        .getMessages().voidLostItems
                        .replace("%prefix%", IridiumSkyblock.getInstance().getConfiguration().prefix)
                        .replace("%items%", String.join(", ", Arrays
                                .stream(lostItemsMap)
                                .map(item -> IridiumSkyblock.getInstance().getMessages().itemsString
                                        .replace("%amount%", item.getAmount())
                                        .replace("%item_name%", item.getItemMeta().hasDisplayName() ? item.getItemMeta().getDisplayName() : "%type%")
                                        .replace("%type%", item.getType().name().trim().replace("_", " ")))
                                .collect(Collectors.toList())))
                ));
```

File: src/main/java/com/iridium/iridiumskyblock/database/LostItems.java (hunk 3 above)
Comment (Faun471) [reply src/main/java/com/iridium/iridiumskyblock/database/LostItems.java:33]: Yeah, me and woof tested this. Here are some of the results:

Player message:
image

Console:
`[23:03:39 INFO]: [IridiumSkyblock] Lost items: Item: LIGHT_BLUE_WOOL, Amount: 20, Name: , Item: YELLOW_WOOL, Amount: 1, Name: , Item: RED_WOOL, Amount: 6, Name: , Item: PURPLE_WOOL, Amount: 5, Name: , Item: LIME_WOOL, Amount: 5, Name:`

File: src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java (hunk 6 above)
Comment (Faun471) [reply src/main/java/com/iridium/iridiumskyblock/listeners/PlayerMoveListener.java]: implemented this in a different commit, but I did change it up a bit, so please take a looksie

**Review by PeachesMLG — APPROVED** [review]

**Pull request merged**